api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -129,6 +156,30 @@
      */
     <T> Optional<T> getOptionalValue(String propertyName, Class<T> propertyType);
 
+    /**
+     * Return the resolved property values with the specified type for the
+     * specified property name from the underlying {@linkplain ConfigSource configuration sources}.
+     * <p>
+     * The configuration values are not guaranteed to be cached by the implementation, and may be expensive
+     * to compute; therefore, if the returned values are intended to be frequently used, callers should consider storing
+     * rather than recomputing them.
+     *
+     * @param <T>
+     *             The property type
+     * @param propertyName
+     *             The configuration property name
+     * @param propertyType
+     *             The type into which the resolved property values should be converted
+     * @return The resolved property values as an {@code Optional} wrapping a list of the requested type
+     *
+     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type
+     */
+    default <T> Optional<List<T>> getOptionalValues(String propertyName, Class<T> propertyType) {
```

Comment:
I believe we must clarify https://github.com/eclipse/microprofile-config/issues/446 first before exposing a return type of `Optional<List<T>>`. I'm usually not a big fan of exposing containers inside containers.

Reply:
`Optional<T[]>` is already possible today, so I don't think this should be blocked by #446.

Reply:
What if we drop these specific multi value methods and introduce a GenericType as an argument, like you have with `readEntity` in JAX-RS?

Then we could just expose a `getValue/getOptionalValue(String propertyName, GenericType type) and you can call it like this:

`.getValue("something", new GenericType<List<String>>() {});`

This will allow you to just overload the methods, instead of adding new ones with the same method parameters and support all the other collection types, instead of just List.

Reply:
Is that better than a SmallRye-style `getValue("something", new ListConverter<String>())`?  Anonymous class type tokens increase metaspace usage...

Reply:
Is that better than a SmallRye-style `getValue("something", new ListConverter<String>())`?  Anonymous class type tokens increase metaspace usage...

Reply:
No, but in that case you are wrapping your `List` in the Converter type and works fine in the method signature with `.getValue(String, Converter)`.

The proposed method is a `.getValues(String, Class)`. First, since the signature is already take, we need to rename the method and then, because we are not able to pass a Collection type, we are always limited by the return type.

I'm fine if we expose a `<T> T getValue(String, Converter<T>)`. This would cover any collection type if you provider a Converter for it. Not so sure about a `<T> List<T> getValues(String, Class<T>)` or the Optional variant.

Reply:
Or maybe we need to think about this in a more integrated way, instead of looking into each issue separately.

Maybe if we implement the ConfigAccessor feature we won't need to add this here and the ConfigAccessor would be responsible to handle all of these cases. Something like, `ConfigAccessor#getValueAsList`?

Reply:
Yeah no answer seems ideal.  Maybe delaying until we have accessors is better.  Or maybe we should just add a documentation line like:

```
If you want a {@code List} of the given type, use: {@code Arrays.asList(config.getValue(key, Item[].class))}.
```

Or something like that.

Reply:
The type token idea has some merit though, I don't mean to dismiss it.  Once nice thing about it is that it gives you parity with CDI.  If I have registered a converter which implements `Converter<Thing<Foo, Bar<Baz>>`, then I could theoretically inject a `Thing<Foo, Bar<Baz>>` in to CDI, but getting such a value using the API is more difficult.  Using the type token approach makes this possible to do with the API.

Reply:
I confess that I didn't thought about that use case. Might be interesting as well, indeed :)

The idea was mostly to make the return type more flexible, instead of being locked down in List.

Reply:
We should really have this discussion in the issue, not the PR, and we _really_ should have had it before I updated the TCK test case 😛

I can certainly sympathize with the issues the name collision may cause with the SmallRye implementation, but didn't we already basically agree that exposing `List<T>` made the most sense?

I'm all for coming up with a more flexible way to get access to multivalued variables, but I think it would be a shame to release 2.0 without an obvious method signature that exposes multivalued properties. The `T[].class` mechanism is really clunky, IMHO.

Reply:
Sorry, I guess we didn't thought well about this :(

The point here is that we need to be extra careful with that we put in there. We could break compatibility in the future if we are not happy with how it turned out, but I believe that once we code something here it will stay forever.

Reply:
At this rate, we won't have to worry about compatibility because we will never change anything.

The premise of micro profile was to move fast and not be paralyzed by "big spec" thinking. We should not be afraid to throw this out when/if accessors come along.

The fact that SmallRye has so much more functionality in SmallRyeConfig than the spec is surely a sign that we're not moving fast enough here.

As an individual contributor in my spare time, I don't want to waste effort on something we're not aligned on. At this point, I would respectfully ask that the committers decide whether to accept or reject this PR so we can move forward.

Reply:
I'm OK with this change as it stands; I think we can improve things later without introducing any backwards compatibility issues.  TBH I don't see a problem with having many useful methods and overloads on this interface.  I think that if you can do more with less, that's great, but it doesn't cost much to do more with more either.  If we do introduce type tokens, these default methods can be efficiently implemented in terms of that solution (as can `getOptionalValue` for that matter).  It really doesn't matter that much.

So my ✅review stands: I'm happy with this as it is.  Even in the absolute worst case where we come up with something better immediately, it doesn't cost us much to have these methods - we already have far more than these in SmallRye Config!

Reply:
Well, I'm sorry. My intention was not to block anything. My concern here was about having a solid and generic API. For that, we don't need to use that much time or be paralysed.

One thing is what SmallRye does on its own. Another thing is what the API does. And BTW, SmallRye does expose it this way:

`public <T, C extends Collection<T>> Optional<C> getOptionalValues(String name, Class<T> itemClass, IntFunction<C> collectionFactory)`

At least, you are not constrained to use only List like the proposal and you can even pass in your own constructed collection. Might as well use it this way.

Reply:
@radcortez this was discussed [in the issue](https://github.com/eclipse/microprofile-config/issues/496#issuecomment-575898716). I'm not necessarily opposed to the SmallRye approach, but I would argue that `List<T>` is the most common/obvious usage. Per @dmlloyd's comments, there's no reason we can't eventually have both. Honestly, I'd say we should add both right now, but I think we'll get pushback in the next hangout 😛

Reply:
If everyone else is happy with the change, I'm not going to push it back, so please move forward.